File: 1339.分裂二叉树的最大乘积.js
/*
 * @lc app=leetcode.cn id=1339 lang=javascript
 *
 * [1339] 分裂二叉树的最大乘积
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxProduct = function(root) {

    const MOD = 1000000007n;

    let total = 0n;
    let max = 0n;

    // 求总和
    function getSum(node) {

        if (!node) {
            return 0n;
        }

        return BigInt(node.val)
            + getSum(node.left)
            + getSum(node.right);
    }

    total = getSum(root);

    // 后序遍历
    function dfs(node) {

        if (!node) {
            return 0n;
        }

        let left = dfs(node.left);
        let right = dfs(node.right);

        let sum = BigInt(node.val) + left + right;

        // 乘积
        let product = sum * (total - sum);

        // 更新最大值
        if (product > max) {
            max = product;
        }

        return sum;
    }

    dfs(root);

    return Number(max % MOD);
};
// @lc code=end

File: 1339.分裂二叉树的最大乘积.java
/*
 * @lc app=leetcode.cn id=1339 lang=java
 *
 * [1339] 分裂二叉树的最大乘积
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    long max = 0;
    long total = 0;
    int MOD = 1000000007;

    public int maxProduct(TreeNode root) {

        // 1. 先求整棵树总和
        total = getSum(root);

        // 2. 再次遍历，计算最大乘积
        dfs(root);

        return (int)(max % MOD);
    }

    // 求整棵树总和
    private long getSum(TreeNode root) {

        if (root == null) {
            return 0;
        }

        return root.val + getSum(root.left) + getSum(root.right);
    }

    // 后序遍历求子树和
    private long dfs(TreeNode root) {

        if (root == null) {
            return 0;
        }

        // 左子树和
        long left = dfs(root.left);

        // 右子树和
        long right = dfs(root.right);

        // 当前子树总和
        long sum = root.val + left + right;

        // 删除当前节点与父节点的边
        // 得到两部分：
        // sum 和 total-sum
        long product = sum * (total - sum);

        // 更新最大值
        max = Math.max(max, product);

        return sum;
    }
}
// @lc code=end

